pylsp/_utils.py

```diff
@@ -220,3 +223,13 @@ def is_process_alive(pid):
             return e.errno == errno.EPERM
         else:
             return True
+
+
+def get_eol_chars(text):
+    """Get EOL chars used in text."""
+    for eol_chars in EOL_CHARS:
+        if text.find(eol_chars) > -1:
+            break
+    else:
+        return None
+    return eol_chars
```

Comment:
The current implementation will be fast for Windows-style file endings (`len(first line)` comparisons), but on Linux it will be `2 * len(entire file) + len(first line)`; a lot of software uses Linux-style file endings. We can get approx `3 * len(first line)` comparisons using regular expressions:

```suggestion
import re

EOL_REGEX = re.compile(f'({"|".join(EOL_CHARS)})')

def get_eol_chars(text):
    """Get EOL chars used in text."""
    match = EOL_REGEX.search(text)
    if match:
        return match.group(0)
    return None
```

A real-world benchmark on 58 of my Python files form one project:

![image](https://user-images.githubusercontent.com/5832902/148647593-889bbcd3-ea48-47df-8cab-d69bc6957673.png)

The regex method is always faster for Linux-style EOL, even for the smallest file which had 40 characters (just one line):

![image](https://user-images.githubusercontent.com/5832902/148647622-75a50ac4-bb9a-405e-85fa-b6fdd60e67ae.png)

For Windows-style file endings it is slower by factor of 2.37, but it's just a constant:

![image](https://user-images.githubusercontent.com/5832902/148648008-c3d07d0c-2672-448f-ad23-8c02872d1bc4.png)

<details>

```python
import timeit
from pathlib import Path
from tqdm.auto import tqdm
from pandas import DataFrame

results = []
for path in tqdm(list(Path('.').glob('*.py'))):
    text = path.read_text()
    if not text:
        continue
    for file_ending in EOL_CHARS:
        text = text.splitlines()
        text = file_ending.join(text)
        old = timeit.timeit('get_eol_chars(text)', globals=globals(), number=10000)
        new = timeit.timeit('get_eol_chars_fast(text)', globals=globals(), number=10000)
        assert get_eol_chars_fast(text) == get_eol_chars(text)
        conditions = {
            'total_len': len(text),
            'first_line_len': len(text.splitlines()[0]),
            'file_ending': repr(file_ending)
        }
        results.append({
            **conditions,
            'time': old,
            'method': 'current'
        })
        results.append({
            **conditions,
            'time': new,
            'method': 'regexp'
        })
df = DataFrame(results)
df.head()
```

</details>

Reply:
(of course the it's still just a fraction of ms even on bigger files, but I have a very fast processor, and it will start becoming noticeable once someone will try to reformat _all_ their files in a larger project, or once this function gets used for other capabilities).

Reply:
Hey @krassowski, thanks for the suggestion! Looks good to me.